### DISCUSSION TITLE
How SDK interact with API

### DISCUSSION THREAD

**Discussion by weizhongzou** [C-Accel-CRIPT/Python-SDK#176] [Q&A]
What would be the cache and de-cache strategy for SDK interacting with API? I do see a lot of cache, i.e., the request made to API for specific data, but what would happen after such request complete? Would the request data stays open, or it would close automatically? I assume those requests are like "fopen", and do we need to have commends like "fclose" to properly "decache" the data stored at API.

**Comment by InnocentBug** [comment]
we do not cache any data from the API other then the DB_schema.
Nodes or other objects are not cached by the SDK.

**Comment by InnocentBug** [comment]
No, `request` does not operate like `fopen` they are complete after the call.
All data is transferred upon completion of the function call.

**Comment by nh916** [comment]
## HTTP Requests

I think this has more to do with [HTTP](https://developer.mozilla.org/en-US/docs/Web/HTTP). 

HTTP is very simple/bare tool and we utilize it to send [GET](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET), [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST), or [Patch](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH) requests to the API.

> HTTP is a [stateless protocol](https://en.wikipedia.org/wiki/Stateless_protocol) and does not remember the past request that you just made to it. Every request to it is new and you have to tell it every time who you are again.

With the [GET](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET) requests we mainly utilize [URL parameters](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams). With [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST) and [Patch](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH) requests we utilize JSON to send data to the API.

We utilize a [JWT token](https://jwt.io/introduction) issued by the API and signed with its cryptographic signature, placed in the JSON header of the JSON request to prove that this request is indeed authenticated and not just anyone unauthorized making requests to our API.

After the API processes the data it then responds back with some [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status), hopefully [200 ](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200) if everything went well and then the JSON representation of what was saved to the database.

If something went wrong, then the HTTP status code would give us some clues as to what is the issue and the API will give an error message as well to help us debug the issue with the HTTP request we sent it (mainly what was wrong with out JSON body or any of the headers that we have improperly configured).

an API error could look like this:
```json
}
    "code": 409,
    "data": null,
    "error": "duplicate item [{'name': 'test_integration_reference_name_068e6295ce8746edb1a4b87f507145b5'}] for Project"
}
```

## Packages SDK Utilizes for HTTP Requests
To send HTTP requests to the API, and then get its response and use it we are using the very popular and well made [requests library](https://requests.readthedocs.io/en/latest/)

> there is a [wiki page about choosing packages](https://github.com/C-Accel-CRIPT/Python-SDK/wiki/How-to-Choose-a-Package), that could help why we chose the specific packages that we did and what we are looking for when choosing a package

requests is pretty much the default package that almost all projects use for sending and receiving HTTP requests

## Resources
Here are some possible resources that might be helpful
* [YouTube video crash course of HTTP protocol](https://youtu.be/iYM2zFP3Zn0)
* [YouTube video crash course of REST API and HTTP protocol](https://youtu.be/Q-BpqyOT3a8)
* [YouTube video on JWT](https://youtu.be/7Q17ubqLfaM)
* [JWT official website](https://jwt.io/introduction)
* [JWT artical by Auth0](https://auth0.com/docs/secure/tokens/json-web-tokens)
